Source/GSAutoLayoutVFLParser.m

```diff
@@ -0,0 +1,756 @@
+/* Copyright (C) 2022 Free Software Foundation, Inc.
+   
+   By: Benjamin Johnson
+   Date: 11-11-2022
+   This file is part of the GNUstep Library.
+   
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+   
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+   
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; if not, write to the Free
+   Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110 USA.
+*/
+
+#import "GSAutoLayoutVFLParser.h"
+#import <AppKit/AppKit.h>
+
+struct GSObjectOfPredicate
+{
+  NSNumber *priority;
+  NSView *view;
+  NSLayoutRelation relation;
+  CGFloat constant;
+};
+typedef struct GSObjectOfPredicate GSObjectOfPredicate;
+
+NSInteger const GS_DEFAULT_VIEW_SPACING = 8;
+NSInteger const GS_DEFAULT_SUPERVIEW_SPACING = 20;
+
+@implementation GSAutoLayoutVFLParser
+
+- (instancetype) initWithFormat:(NSString *)format
+                       options:(NSLayoutFormatOptions)options
+                       metrics:(NSDictionary *)metrics
+                         views:(NSDictionary *)views
+{
+  if (self = [super init])
+    {
+      if ([format length] == 0)
+        {
+          [self failParseWithMessage:@"Cannot parse an empty string"];
+        }
+
+      _views = views;
+      _metrics = metrics;
+      _options = options;
+
+      _scanner = [NSScanner scannerWithString:format];
```

Comment:
All these are autoreleased object, they could go away when passed over the border of an autorelease pool. Most likely this is no issue here as this object will be short lived.